tool2schema/schema.py

```diff
@@ -75,6 +75,74 @@ def SaveGPTEnabled(module: ModuleType, path: str, schema_type: SchemaType = Sche
     json.dump(schemas, open(path, "w"))
 
 
+class ParseException(Exception):
+    """Exception for schema parsing errors."""
+
+    pass
+
+
+def ParseSchema(module: ModuleType, function, validate: bool = True) -> Optional[Callable]:
+    """
+    Given a function object (`openai.types.chat.chat_completion_message_tool_call.Function`)
+    retrieve the corresponding function among those with the `GPTEnabled` decorator defined in
+    `module`. When `validate` is true, validate the arguments string and raise `ParseException`
+    if the arguments are not valid and cannot be used to invoke the function.
+
+    :param module: The module where the function is defined
+    :param function: An instance of OpenAI Function model
+    :param validate: Whether to validate the function arguments
+    :return: The function defined in `module`, or None if no function with the expected name
+        and the `GPTEnabled` decorator is found
+    :raises ParseException: If the argument string is not valid, meaning it is not parsable as JSON,
+        it can be parsed but is not a dictionary, it does not include all required arguments, it
+        includes values that are not of the expected type, or it contains hallucinated arguments
+    """
+
+    arguments = None
+
+    if validate:
+        try:
+            arguments = json.loads(function.arguments)
+
+        except json.decoder.JSONDecodeError:
+            raise ParseException("Arguments are not in valid JSON format")
+
+        if type(arguments) is not dict:
+            raise ParseException("Arguments are not in the form of a dictionary")
+
+    f = FindGPTEnabledByName(module, function.name)
+
+    if f and validate:
+        _validate_arguments(f, arguments)
+
+    return f
```

Comment:
I think we should return a lambda that you can call to call `f` with the arguments loaded already.

In other words:

```
return lambda: f(**arguments)
```

Note sure if the above works but you get the idea.

Reply:
What about returning a tuple `(f, args)`, which you invoke as `f(**args)`? This would not 'hide' the parsed arguments dictionary from the caller

Reply:
This also allows for better testing
```python
f, args = ParseSchema(functions, f_dict)
assert f == function
assert args == arguments # Verify the hallucinated argument has been removed
```

Reply:
Make a comment about the `None` return below. Otherwise, I think this is fine.